### PR TITLE
Add possibility to user defined custom task verdict producers

### PR DESCRIPTION
**Pull request by sergioboi**

### CODE REVIEW

File: subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/internal/BuildExecutionFailureListener.kt [hunk 1]
```diff
@@ -29,7 +30,8 @@ internal class BuildExecutionFailureListener(
                     FailedTask(
                         name = task.name,
                         projectPath = task.project.path,
-                        errorOutput = logs[Path.path(task.path)]?.build() ?: "No error logs",
+                        errorLogs = logs[Path.path(task.path)]?.build() ?: "No error logs",
```
Comment (sergioboi) [inline subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/internal/BuildExecutionFailureListener.kt:33]: Separate logs and verdict

File: subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/internal/task/lifecycle/UserDefinedVerdictProducerTaskLifecycleListener.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.avito.android.build_verdict.internal.task.lifecycle
+
+import com.avito.android.build_verdict.UserDefinedTaskVerdictProducer
+import com.avito.android.build_verdict.internal.LogsTextBuilder
+import org.gradle.api.Task
+import org.gradle.util.Path
+
+internal class UserDefinedVerdictProducerTaskLifecycleListener(
```
Comment (sergioboi) [inline subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/internal/task/lifecycle/UserDefinedVerdictProducerTaskLifecycleListener.kt:8]: 🎉

File: subprojects/gradle/build-verdict/src/test/kotlin/com/avito/android/build_verdict/PlainTextVerdicts.kt [hunk 3]
```diff
@@ -18,10 +18,13 @@ fun unitTestsFails(dir: File) = """
 Execution failed for task ':app:testDebugUnitTest'.
 	> There were failing tests. See the report at: file://${dir.canonicalPath}/app/build/reports/tests/testDebugUnitTest/index.html
 
-* Error logs:
+* Task result:
```
Comment (sergioboi) [inline subprojects/gradle/build-verdict/src/test/kotlin/com/avito/android/build_verdict/PlainTextVerdicts.kt:21]: That's how the verdict changed

File: subprojects/gradle/ci-logger/src/main/kotlin/com/avito/utils/logging/CILoggingDestination.kt [hunk 4]
```diff
@@ -9,6 +9,16 @@ interface CILoggingDestination : Serializable {
     fun child(tag: String): CILoggingDestination
 }
 
+internal object StderrDestination : CILoggingDestination {
```
Comment (sergioboi) [inline subprojects/gradle/ci-logger/src/main/kotlin/com/avito/utils/logging/CILoggingDestination.kt:12]: The cheapest way to help BuildVerdictPlugin see error logs from our tasks

File: subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildVerdictPluginExtension.kt [hunk 5]
```diff
@@ -12,4 +14,32 @@ abstract class BuildVerdictPluginExtension(
 ) {
     val outputDir: Property<Directory> =
         objects.directoryProperty().convention(layout.projectDirectory.dir("outputs/build-verdict"))
+
+    val taskVerdictProviders = objects.listProperty<UserDefinedTaskVerdictProducer>()
```
Comment (dsvoronin) [inline subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/BuildVerdictPluginExtension.kt]: can you hide it with `internal`?
as far as i understand client supposed to use functions below?

File: subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/UserDefinedTaskVerdictProducer.kt [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    class ByClass(private val acceptedClass: Class<in Task>) : TaskPredicate {
+        override fun accept(task: Task) = acceptedClass.isInstance(task)
+    }
+}
+
+interface TaskVerdictProducer {
+    fun produce(task: Task): String
+
+    class Lambda(private val producer: (Task) -> String) : TaskVerdictProducer {
```
Comment (dsvoronin) [inline subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/UserDefinedTaskVerdictProducer.kt]: I see no point in this additional class, doesn't 

```
companion object {
   internal fun create(producer: (Task) -> String) = object : TaskVerdictProducer {
      override fun produce(task: Task) = producer(task)
   }
}
```

tells intention the easier way?
I see this lambda as some synthetic thing

Comment (dsvoronin) [inline subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/UserDefinedTaskVerdictProducer.kt]: could be internal

File: subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/UserDefinedTaskVerdictProducer.kt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+) : TaskPredicate by predicate, TaskVerdictProducer by producer
+
+interface TaskPredicate {
+    fun accept(task: Task): Boolean
+
+    class ByName(private val name: String) : TaskPredicate {
+        override fun accept(task: Task) = task.name == name
+    }
+
+    class ByClass(private val acceptedClass: Class<in Task>) : TaskPredicate {
```
Comment (dsvoronin) [inline subprojects/gradle/build-verdict/src/main/kotlin/com/avito/android/build_verdict/UserDefinedTaskVerdictProducer.kt]: more like ByType? 

examples: https://github.com/gradle/gradle/search?q=withType
some thoughts: https://stackoverflow.com/a/16600781/981330

**Review by dsvoronin — COMMENTED** [review]
I really like how this plugin shaped, good work!

**Review by eugene-krivobokov — APPROVED** [review]

**Pull request merged**